engine/access/state_stream/backend/handler.go

```diff
@@ -134,71 +154,195 @@ func (h *Handler) SubscribeEvents(request *executiondata.SubscribeEventsRequest,
+}
+
+// handleEventsResponse handles the event subscription and sends subscribed events to the client via the provided stream.
+// This function is designed to be used as a callback for events updates in a subscription.
+// It takes a EventsResponse, processes it, and sends the corresponding response to the client using the provided send function.
+//
+// Parameters:
+// - send: The function responsible for sending events response to the client.
+//
+// Returns a function that can be used as a callback for events updates.
+//
+// Expected errors during normal operation:
+//   - codes.Internal - could not convert events to entity or the stream could not send a response.
+func (h *Handler) handleEventsResponse(send sendSubscribeEventsResponseFunc, requestHeartbeatInterval uint64, eventEncodingVersion entities.EventEncodingVersion) func(*SubscribeEventsResponse) error {
+	heartbeatInterval := requestHeartbeatInterval
 	if heartbeatInterval == 0 {
 		heartbeatInterval = h.defaultHeartbeatInterval
 	}
 
 	blocksSinceLastMessage := uint64(0)
-	for {
-		v, ok := <-sub.Channel()
-		if !ok {
-			if sub.Err() != nil {
-				return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
-			}
-			return nil
-		}
-
-		resp, ok := v.(*EventsResponse)
-		if !ok {
-			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
-		}
 
+	return func(resp *SubscribeEventsResponse) error {
 		// check if there are any events in the response. if not, do not send a message unless the last
 		// response was more than HeartbeatInterval blocks ago
 		if len(resp.Events) == 0 {
 			blocksSinceLastMessage++
 			if blocksSinceLastMessage < heartbeatInterval {
-				continue
+				return nil
 			}
 			blocksSinceLastMessage = 0
 		}
 
 		// BlockExecutionData contains CCF encoded events, and the Access API returns JSON-CDC events.
 		// convert event payload formats.
 		// This is a temporary solution until the Access API supports specifying the encoding in the request
-		events, err := convert.EventsToMessagesWithEncodingConversion(resp.Events, entities.EventEncodingVersion_CCF_V0, request.GetEventEncodingVersion())
+		events, err := convert.EventsToMessagesWithEncodingConversion(resp.Events, entities.EventEncodingVersion_CCF_V0, eventEncodingVersion)
 		if err != nil {
 			return status.Errorf(codes.Internal, "could not convert events to entity: %v", err)
 		}
 
-		err = stream.Send(&executiondata.SubscribeEventsResponse{
-			BlockHeight: resp.Height,
-			BlockId:     convert.IdentifierToMessage(resp.BlockID),
-			Events:      events,
+		err = send(&executiondata.SubscribeEventsResponse{
+			BlockHeight:    resp.Height,
+			BlockId:        convert.IdentifierToMessage(resp.BlockID),
+			Events:         events,
+			BlockTimestamp: timestamppb.New(resp.BlockTimestamp),
+			MessageIndex:   resp.MessageIndex,
```

Comment:
similar to the account statuses endpoint, we'll need to move the `MessageIndex` tracking here since empty messages may be dropped. See https://github.com/onflow/flow-go/pull/5406#discussion_r1543838354

Reply:
[de78346](https://github.com/onflow/flow-go/pull/5602/commits/de78346b3a43d4d9b975b3e9611cfec41dd9e170)